Natalie

natpol100 opened the pull request:
I've created: 
01_combine_data.R 
01_combine_data_function.R 

I've also cut and pasted the library commands for these scripts into 00_setup_environment

Could you review these scripts and give me some feedback?

Commit:
Added libraries for 01_combine_data.R to the 00_setup_environment.R script.

Modified the 01_combine_date.R script

Commit:
Update to 01_combine data.R

Add first version of 01_combine_data_function.R to the functions folder

Commit:
Update to 01_combine data.R

Add first version of 01_combine_data_function.R to the functions folder.

path_3 has been edited so that the file for Discovery that includes the date in the file name is saved directly into the archive folder.

path_4 has been added

Commit:
First draft of 01_combine_board_narrative

Commit:
Deleted decorators resulting from merge conflict

Merge branch 'natalie' of https://github.com/Public-Health-Scotland/chronic-pain into natalie

Conflicts:
	functions/01_combine_data_function.R

Commit:
Edit to remove repetition of code caused by merge conflict

Commit:
Second edit to remove repetition of code related to path_4 caused by earlier merge conflict

Commit:
Update to list of folders where output files will be saved: \\nssstats01.csa.scot.nhs.uk\WaitingTimes\Chronic-Pain\Discovery\archive has been added to the list

Commit:
Added stringr package.
Removed commented out lines.
Edited cell range being read from submissions.
Now using str_remove_all to remove file extensions

rmccreath commented on code/00_setup_environment.R:
```diff
@@ -0,0 +1,18 @@
+###                                                                                    ###
+###   Description: This script sets up the environment                                 ###
+###                                                                                    ###
+##########################################################################################
```
This is more of a personal preference/taste thing but you could remove the last column of '#'s which will make changes easier to manage (you don't have to line them up every time you change something on that line). Also, you can just start each line with 1 `#`.

rmccreath commented on code/00_setup_environment.R:
```diff
@@ -0,0 +1,18 @@
+###                                                                                    ###
+##########################################################################################
+
+###   Load libraries required   ###
```
Every script should be structured with sections. We'll be able to use this script for more later, so further sections will likely be added. 

```suggestion
### 1 - Load libraries and functions required ----
```
I use this format for the section headers `### x - title ----` - the four dashes allow you to "collapse" code in RStudio which makes it a bit easier to navigate while developing (it's also still clean enough to not distract).

rmccreath commented on code/00_setup_environment.R:
```diff
@@ -0,0 +1,18 @@
+library(readxl)
+library(stringr)
+library(tidyr)
+library(writexl)
```
Not entirely necessary, but a comment after each package with a brief description of why that package is used will be useful for maintenance. 

e.g.
```
library(dplyr)          # For data manipulation in the "tidy" way
```

rmccreath replied:
Another line can be added that will read all functions from the `/functions` directory. This will require `purrr` be added to the list of packages.

```
library(purrr)          # For iterative functional programming with vectors

walk(list.files(here("functions"), full.names = TRUE), source)
```

rmccreath replied:
Other packages which will either already be in use or come in useful (probably not exhaustive):

```
library(readr)        # For reading files
library(lubridate)    # For dealing with dates
library(janitor)      # For 'cleaning' variable names
library(phsmethods)   # For internal PHS functions
library(magrittr)     # For the %<>%
library(ggplot2)      # For creating charts
library(knitr)        # For creating tables in markdown
library(rmarkdown)    # For rendering markdown documents
library(zip)          # For archiving files
library(here)         # For the here() function
```

rmccreath commented on code/00_setup_environment.R:
```diff
@@ -0,0 +1,18 @@
+
+###   Load libraries required   ###
+library(dplyr)
+library(readxl)
```
Rather than `readxl`, the recommended package for dealing with Excel files is `openxlsx`. I understand this would require some deeper work than just changing the package but it'll be good to add it now and work towards changing the functions as you're developing them. (The same is true for `writexl`).

rmccreath commented on code/01_board_narrative.R:
```diff
@@ -0,0 +1,75 @@
+###   1. Get file names   ###
+
+# Set working directory
+setwd("//PHI_conf/WaitingTimes/Chronic-Pain/Data/Submissions & emails/Current Quarter/")
```
I think we need to work toward having the data files be included as part of the RAP project, a (temporary) workaround could be _copying_ the external files as part of the `setup_environment` script but we can't have these static file-paths throughout the code or be changing the working directory.

rmccreath commented on code/01_board_narrative.R:
```diff
@@ -0,0 +1,75 @@
+path_tables = "//PHI_conf/WaitingTimes/Chronic-Pain/Publications/Outputs/Narrative_xxxxxxxx.xlsx"
+
+# Write Excel file
+write_xlsx(notes, path_tables)
```
Keep all the project outputs within the project, so using `here::here()`, write out to `/data/output`. Also, let's not write out Excel files if we don't need to. We can export this data-frame as RDS (`write_RDS`) which is more efficient.

rmccreath commented on code/01_board_narrative.R:
```diff
@@ -0,0 +1,75 @@
+# Initiate a blank data frame
+notes <- data.frame()
+
+for (file in filenames){
```
As this is quite small and related to the `combine_data` script, we can make a `create_notes` function to do this and combine the scripts? Also, can then use `purrr::map`.

rmccreath commented on code/01_board_narrative.R:
```diff
@@ -0,0 +1,75 @@
+
+# Write Excel file
+write_xlsx(notes, path_tables)
+
```
Just as a note, where you do `variable <- variable %>% code` you can write `variable %<>% code` instead.

rmccreath commented on code/01_combine_data.R:
```diff
@@ -0,0 +1,35 @@
+###   NOTE: Update lines 32 - 34 BEFORE running the function                           ###
+###   Line 32: Input last quarter end date):  e.g. "2021-03-31"                        ###                                                 
+###   Line 33: Input current quarter end date): e.g. "2021-06-30"                      ###                                                          
+###   Line 34: Input current quarter end date): e.g. "2021-06-30"                      ###                                              
```
These variables should be calculated automatically or where not possible manually updated variables should be in the `setup_environment` script. 

```
### 0 - Manual Variable(s) - TO UPDATE ----

# UPDATE - First day of reporting quarter (ddmmyyyy)
current_qtr_end <- lubridate::dmy(30062021)


### 1 - Load libraries and functions required ----
[...]


### 2 - Derive date variables
previous_qtr_end <- as_date(cut(dmy(current_qtr_end), "quarter")) - 1
```

rmccreath commented on code/01_combine_data.R:
```diff
@@ -0,0 +1,35 @@
+##########################################################################################
+
+### Source the function for combining the most recent versions of ALLDATA and Current Quarter 
+source("//PHI_conf/WaitingTimes/Chronic-Pain/RAP/chronic-pain-main/functions/01_combine_data_function.R")
```
All functions are now read in as part of the `setup_environment` script.

rmccreath commented on functions/01_combine_data_function.R:
```diff
@@ -0,0 +1,102 @@
+##########################################################################################
+
+### Function for combining ALLDATA and the Current quarter files
+Combine_ALLDATA_Currentquarter <- 
```
Function names should be all lowercase and the filename should also just be the name of the function (not the script it gets run in, that's one of the reasons we use functions - they're applicable outside of a single-use).

rmccreath commented on functions/01_combine_data_function.R:
```diff
@@ -0,0 +1,102 @@
+
+### Function for combining ALLDATA and the Current quarter files
+Combine_ALLDATA_Currentquarter <- 
+  function(date_existing_ALLDATA, date_currentquarter,date_new_ALLDATA)
```
Similar to the name of functions, the name of arguments should be all lowercase. Also, doesn't `date_currentquarter` and `date_new_ALLDATA` correspond to the same date? If so, these should be condensed.

rmccreath commented on functions/01_combine_data_function.R:
```diff
@@ -0,0 +1,102 @@
+    
+    # Write Excel file for Discovery with required naming convention
+    # Will overwrite existing file
+    write_xlsx(Discovery, paste0(path_4, "Chronic Pain WT - All Data.xlsx"))
```
Why is the data getting written to archives at the same time it's getting saved to the main data location? This should probably be done as a check at the beginning of the project process rather than doubling process times at this stage?

rmccreath reviewed (changes requested):
Hey, I've tried to be specific and provide reasoning for any changes, hope this is helpful. Happy to discuss anything or help in another way if needs be. I really like your comments at the start of each script (inputs and outputs), this is really helpful.

Overall, there's a lot of static file paths through the code which need to be minimised. I understand it won't be possible everywhere but we should review the overall process to make sure. Related to this, I think we need to copy in the data at the start of the project and mainly work from there, with the view to incorporate the data management as part of the RAP process?

Have a look at the style guidance (https://github.com/Public-Health-Scotland/R-Resources/blob/master/PHS%20R%20style%20guide.md). There's a couple of things that will make the code easier to read and maintain (variable names, line length, script structure, comments, etc.).

Commit:
Updates made to 00_setup_environment.R based on feedback from Russell

Commit:
Added ### END OF SCRIPT ###

Commit:
Made some of changes suggested by Russell

Commit:
Updates to 01_board_narrative

Commit:
Added packages to list and tidied up script

Commit:
Added line to input end date of current quarter and line to derive date of previous quarter

Commit:
Update to header

Commit:
Deleted file 01_combine_data.R

Commit:
Renamed file

Commit:
Added ### END OF SCRIPT ###

Commit:
Added section numbering

Commit:
Renamed script as combine_narratives and turned it into a function

Commit:
Added section numbering

Commit:
Added 4 dashes after section 2 header to be able to collapse it

Commit:
Added paths to 00_setup

Commit:
Added paths and code for copying in files.
Added some questions about how to read in 14 submissions

Commit:
Changed files paths so that data is read in from within input folder and 2 files are saved into output folder

Commit:
Changed file name

Commit:
Added path for reading submissions from data/submissions folder

Commit:
New script for running the read_narrative function over 14 files

Commit:
New script for running the read_submission function over 14 files

Commit:
Added comments to explain what paths are for

Commit:
Added path for reading submission data from data/submissions folder

Commit:
Added path for reading submission data from data/submissions folder

Commit:
Updated comments about files being saved

Commit:
Edited script name and rewrote as a function that can be used by map_dfr

Commit:
New function for reading data from different tabs of the Excel submissions

rmccreath commented on code/00_setup_environment.R:
```diff
@@ -38,12 +38,57 @@ library(openxlsx)     # For reading and writing Excel files
+# Path for copying in lookup file for rate of referral calculation
+path_lookup = "//PHI_conf/WaitingTimes/Chronic-Pain/R development work/Lookups/Populationestimatemid2020_18+.xlsx"
+
+# Path for writing out the Board Narrative column for the workbook
+path_narrative =  here("data", "output")
+
+# Path for writing out current quarter dataframe
+path_currentquarter = here("data", "input")
+
+# Path for writing out updated ALL DATA file
+path_discovery =  here("data", "output")
+
+# Path for writing out file for Discovery
+path_newalldata = here("data", "output")
```
You have 3 paths here that are all the same, abstract these to a single `path_output` variable.

rmccreath commented on code/01_combine_submission.R:
```diff
@@ -0,0 +1,49 @@
+
+### 3 - Write file with current_quarter_end date included in file name  ----
+
+# NEED TO decide if writing current_quarter file is needed longer-term? Or would 
+# current_quarter be created in R and then added to the bottom of the ALL DATA 
+# file without saving current_quarter first?
```
Writing and reading uses a lot of compute. I think the preferable option would be to keep everything within R and attempt to do all the necessary processing. Any outputs should ideally be the final outputs for the publication.

rmccreath commented on code/01_combine_submission.R:
```diff
@@ -0,0 +1,49 @@
+
+write.xlsx(current_quarter, paste0(path_currentquarter, current_qtr_end, ".xlsx"))
+
+# Or should I use write_RDS instead?
+#write_rds(current_quarter, paste0(path_currentquarter, current_qtr_end, ".rds"))
```
If this is still being processed, then using `.RDS` is preferred.

rmccreath commented on functions/combine_data.R:
```diff
@@ -72,32 +66,32 @@ combine_data <-
     
     # Set path for exporting to \\nssstats01.csa.scot.nhs.uk\WaitingTimes\Chronic-Pain\Data\Database
-    path_1 = "//PHI_conf/WaitingTimes/Chronic-Pain/Data/Database/"
+    #path_1 = "//PHI_conf/WaitingTimes/Chronic-Pain/Data/Database/"
```
This is currently set up as a variable, `path_alldata` from the `setup_environment` script. Avoid static file paths as much as possible, and when they're necessary make sure they're not hidden within code. This is true for paths following this too.

rmccreath commented on functions/read_narrative.R:
```diff
@@ -0,0 +1,120 @@
+#         TRUE ~ (as.character(`Board narrative`)) 
+#       ))
+#     
+#     boardname <- data.frame(paste0("NHS ", file)) %>%
+#               rename(`Board narrative` = `paste0..NHS....file.`)
+#     
+#     notes <- rbind(notes, boardname, temp_notes)
+#     rm(boardname)
+#     rm(temp_notes)
+#     
+#     notes %<>% 
+#       mutate(`Board narrative` = str_remove_all(`Board narrative`, "\\.xlsx$"))
+#     
+#     }
+# 
+# 
+#     ### 3 - Write Excel file for the workbook ----
+# 
+#     # Set path for saving Board narrative
+#     #path_narrative =  here("data", "output")
+#     #path_tables =  here("data", "output", "Narrative_xxxxxxxx.xlsx")
+# 
+#     # Write Excel file to output folder
+#     
+#     #write_xlsx(notes, path_tables)
+#     write_xlsx(notes, paste0(path_narrative, current_qtr_end, "_Narrative.xlsx" ))
+#     
+#   }
+# 
+# 
+# ### END OF SCRIPT ###
+# 
+
+
+
+
+
+
+
+
```
This is all included with the git history, you can delete this.

rmccreath reviewed (approved):
Without having the chance to run on actual data, I've only got some minor follow-up comments, all added to the code. I've approved so when those comments are addressed, feel free to merge.

Commit:
Changes made to file path at line 49 and walk function removed from line 73

Commit:
Edited file paths to have one for the output file and one for the input file

Commit:
Edited line 21 to remove walk and source.
Edited file path used in write_xlsx

Commit:
Deleted older version of code that was commented out

Commit:
Added current_qtr_end as a argument to map

Commit:
files which gives the list of file paths has been added as a input to the function.

Path name has been changed to files to match the script that runs the function.

Commit:
Edited files to remove walk and source.

Deleted questions marked as comments.

Deleted write_xlsx code and replaced with write_rds

Commit:
Edited comment

Commit:
Added files as an input to the function.

Removed commented out lines.

Replaced path_readsubmissions with files

Commit:
Deleted commented out code.

Changed code for reading in the current quarter's data from the input file to read in a .rds file.

Deleted commented out paths for saving 4 files now save 2 versions to the output folder - one for Waiting Times and one for Discovery.

Commit:
Removed commented out code.

Changed from a function to a script.

Commit:
Moved from the function folder to the code folder

Commit:
Moved from the functions folder to the code folder and rename as 01_combine_data.R

rmccreath commented on code/00_setup_environment.R:
```diff
@@ -0,0 +1,85 @@
+
+
+# Read in ALL DATA file from previous quarter
+file.copy(paste0(path_alldata, previous_qtr_end, " ALL DATA.xlsx"), here("data", "input"))
```
You set up the `path_input` variable, this is somewhere that you could use it. This happens a few times throughout different scripts though so I'll let you decide how to continue.

rmccreath commented on code/01_combine_narrative.R:
```diff
@@ -0,0 +1,37 @@
+
+files <- list.files(here("data", "submissions"), full.names = TRUE)
+
+# Set path to use in the read_narrative function
+#path_readsubmissions = here("data", "submissions", files, ".xlsx")
```
If this isn't used, delete. If this is sometimes used, create some logic around it to run when necessary.

rmccreath reviewed (approved):
Hey, all of this looks great and nothing to stop it being merged to the main branch. Generally looks to perform as expected and all previous review comments have been addressed. Only a couple of broad best practice / styling things.

Commit:
Changed name to notes to match what is being output by the function and written to the output folder

Commit:
overwrite = TRUE has been added to lines 70, 74 and 78

Commit:
Added file name to line 36

Commit:
Added file name to line 31

Commit:
Sometimes the pain psychology line on the referrals tab contains a row of NAs, so need to read in column 1 and then drop it to ensure that don't end up with an empty df, as this stops cbind() from working

Commit:
Added slashes to end of input and output file paths.

Changed path to files

Commit:
Added html output to gitignore
